File: lc-python/src/lc_0876_middle_linked_list.py
import unittest
from typing import Optional

from shared.linked_list import ListNode, to_list_node, list_node_to_list


def middle_linked_list_recursive(head: Optional[ListNode]) -> Optional[ListNode]:
    """
    Time complexity: O(n)
    Space complexity: O(n)
    """

    def length(node: Optional[ListNode], curr=0) -> int:
        if not node:
            return curr
        return length(node.next, curr + 1)

    def get(node: Optional[ListNode], idx) -> Optional[ListNode]:
        if not node or idx < 0:
            return None
        if idx == 0:
            return node
        return get(node.next, idx - 1)

    return get(head, length(head) // 2)


def middle_linked_list(head: Optional[ListNode]) -> Optional[ListNode]:
    """
    Time complexity: O(n)
    Space complexity: O(1)
    """
    slow = head
    fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
    return slow


class TestMiddleLinkedList(unittest.TestCase):
    @staticmethod
    def parse_input():
        data = [
            [[1, 2, 3, 4, 5], [3, 4, 5]],
            [[1, 2, 3, 4, 5, 6], [4, 5, 6]],
        ]
        return data

    def test_middle_linked_list_recursive(self):
        test_cases = self.parse_input()
        for tc in test_cases:
            xs = to_list_node(tc[0])
            expected = tc[1]

            actual = list_node_to_list(middle_linked_list_recursive(xs))

            self.assertEqual(actual, expected)

    def test_middle_linked_list(self):
        test_cases = self.parse_input()
        for tc in test_cases:
            xs = to_list_node(tc[0])
            expected = tc[1]

            actual = list_node_to_list(middle_linked_list(xs))

            self.assertEqual(actual, expected)


if __name__ == "__main__":
    unittest.main()

File: lc-typescript/src/lc_0876_middle_linked_list.ts
import { assertEquals } from "https://deno.land/std@0.224.0/assert/assert_equals.ts";
import { ListNode, array_to_linked_list, linked_list_to_array } from "./shared/utils.ts";

function middle_linked_list(head: ListNode | null): ListNode | null {
    let slow = head;
    let fast = head?.next || null;
    while (fast != null) {
        slow = slow?.next || null;
        fast = fast.next?.next || null;
    }
    return slow;
}

Deno.test("0876 Middle of Linked List", () => {
    const test_cases: [number[], number[]][] = [
        [[1, 2, 3, 4, 5], [3, 4, 5]],
        [[1, 2, 3, 4, 5, 6], [4, 5, 6]],
    ];

    for (let i = 0; i < test_cases.length; i += 1) {
        const test_case = test_cases[i];

        const head = array_to_linked_list(test_case[0]);
        const expected = test_case[1];

        const actual = linked_list_to_array(middle_linked_list(head));

        assertEquals(actual, expected);
    }
});
